Adding provider hooks

jakewan opened the pull request:
This PR was closed and is superseded by https://github.com/cbsinteractive/open-insights/pull/6.

This is a major reorganization of the core Open Insights module intended to give providers the means of customizing most aspects of measurement functionality.

jakewan committed:
Accumulated changes

jakewan committed:
Remove expletive

jakewan committed:
Accumulated changes

jakewan committed:
Accumulated changes 2020-05-31

jakewan committed:
Accumulated changes 2020-06-01

jakewan committed:
Accumulated changes 2020-06-01

jakewan committed:
Rename TaskInterface to IExecutable

jakewan committed:
Allow provider to specify if and how it obtains fetch headers

jakewan committed:
Remove src/lib/pop.ts

Specialized functionality is handled by the provider

jakewan committed:
Update TypeScript to 3.9.3

jakewan committed:
Begin work adding unit tests

- ported assign tests from insights.js

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,92 @@
+    | boolean
+}
+
+export interface IExecutable {
```
There are two naming conventions for interfaces in this file, `IFoo` prefix and `FooInterface` postfix, most TypeScript style-guides/linters try to avoid the use of either of these patterns. We should at least choose one convention and stick with it. I personally feel both `Executable` and `Provider` don't need any interface notation.

A wider discussion point for the project would be to formalise some style-guide.

jakewan replied:
This makes sense re: Executable and Provider. Maybe we can discuss beginning work on a style-guide and other contribution guidelines when we talk next week.

jakewan replied:
https://github.com/cbsinteractive/open-insights/issues/3

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,92 @@
+
+export interface ClientInfoResponseFunc {
+    (response: Promise<any>): Promise<unknown>
+}
```
Looking at their usage, I don't think either of these need to be interfaces and could be function type literals instead?

```suggestion
export type PromiseSequenceFunc = (sessionConfigs: SessionConfig[]) => Promise<unknown>

export type ClientInfoResponseFunc = ()(response: Promise<any>) =>  Promise<unknown>
```

jakewan replied:
Yes, definitely.

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,92 @@
+export interface FetchResponse {
```
Nit. This file could have been `git mv src/@types/index.d.ts src/@types/index.ts` to keep this history.

jakewan replied:
Interesting. Didn't realize `git mv` had this effect. I always wondered how sometimes git is able to mark a file as renamed rather than deleted, and other times not. I'll try to remember this for the future.

phamann commented on src/index.ts:
```diff
@@ -14,31 +11,27 @@ export function init(settings: PageSettings): void {
-                    }
-                }
+function start(settings: ClientSettings): Promise<any> {
+    return new Promise<any>(resolve => {
```
Why are we constructing a new `Promise` here instead of using the one returned by the `Promise.all()` on the line below? I also don't see the resolve being called anywhere?

jakewan replied:
Good catch. Probably just me being confused. I'll remove the superfluous outer promise.

phamann commented on src/lib/providerbase.ts:
```diff
@@ -0,0 +1,49 @@
+
+    shouldRun(): boolean {
+        return true
+    }
```
Maybe this should be `abstract` to force a provider implementer to consider it?

jakewan replied:
Yes, definitely. Done.

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,41 @@
+            )
+    }
+
+    fetchObjectAndId(): Promise<string> {
+        return fetch(this.getResourceUrl())
+            .then((res): string => {
+                return this.provider.getFetchHeaders(res.headers, this.config)
+            })
+    }
```
`fetchObjectAndId()` is a historic detail to the Fastly provider implementation, specifically returning an ID string from a set of headers and may not not be useful in the core library? Maybe we could return the response object and pass that to `createFetchResult()` on line 30 above? This would allow providers to implement whatever logic they want to acquire an ID instead of just `getFetchHeaders`?
```suggestion
    fetchObject(): Promise<string> {
        return fetch(this.getResourceUrl())
    }
```

jakewan replied:
Coming from Cedexis where this feature was always a big deal, it seemed natural to include in the base implementation, but I think your idea is much nicer.

phamann commented on src/lib/providerbase.ts:
```diff
@@ -0,0 +1,49 @@
+     * A subclass overrides this if it wants to obtain the value of any fetch response headers
+     * @param headers
+     */
+    getFetchHeaders(headers: Headers, testConfig: unknown): string {
```
See comment in `fetch.ts` this may not be needed.

jakewan replied:
Yep, removed this along with the related suggested change.

phamann commented on src/index.ts:
```diff
@@ -14,31 +11,27 @@ export function init(settings: PageSettings): void {
+            return settings.sequence(sessionConfigs)
+        }).then(data => {
+            console.log('Finished!')
+            console.log(data)
```
Assume these were left for testing purposes. We should discuss how we may want to facilitate a debug logging pattern, via a `utils/log.ts` `log()` method or similar, which could be a no-op in production but would log if a config setting is passed etc? 

Similar comments about `console.log` statements are elided for rest of review.

jakewan replied:
I'll remove them now, and let's discuss a better way forward as you suggest.

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+    hasRan: boolean
+}
+
+export abstract class TestBase implements IExecutable {
```
Similar to the interface notation discussion, I'm intrigued to understand the `FooBase` naming convention. I'm not against it. But `Test` seems sufficient to me?

jakewan replied:
A long time ago, it seemed like interfaces and abstract base classes were commonly named in a way to make that clear to everyone. I don't know what it's like now, but back in the 90s, when I was first learning Windows programming, you'd often see interface names prefaced with a capital "I", abstract base classes ending in "Base". You'd sometimes even see class names prefaced with "C"! 😬 

For the last 10ish years, it seems I haven't had much time to stay abreast of modern object-oriented programming. I really enjoy this opportunity to come up to date, including learning about the naming conventions preferred by the community.

Agreed that bases classes should just be named simply, without the "Base" suffix. I'll make that change wherever it's been done.

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export abstract class TestBase implements IExecutable {
+    public state: State = { hasRan: false }
```
This is historic implementation detail from Fastly, I'm happy to leave however we should make sure it is actually set inside `execute()`.

jakewan replied:
I went with this: https://github.com/cbsinteractive/open-insights/commit/41795469a70c805b404b9867f7d5c2da720cda99

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,41 @@
+    test(): Promise<ResourceTimingEntry> {
+        return Promise.all<string, ResourceTimingEntry>([
+            this.fetchObjectAndId(),
+            asyncGetEntry(this.getResourceUrl(), 5000)
```
Note: we recently found a bug in the implementation of `asyncGetEntry` in insights.js so we should ensure this is also patched in the core library: https://github.com/fastly/insights.js/commit/c05c4a239933d53c958753c2da433572a27328f5

jakewan replied:
https://github.com/cbsinteractive/open-insights/commit/9d7735d635541b825d87bd4fd5b67d08990f8148

phamann commented on src/lib/providerbase.ts:
```diff
@@ -0,0 +1,49 @@
+    abstract name: string
+    abstract fetchSessionConfig(): Promise<SessionConfig>
+    abstract expandTasks(): IExecutable[]
+    abstract makeBeaconData(testConfig: unknown, testData: unknown): unknown
```
I feel slightly uncomfortable about making `testData` be `unknown`, as it makes it very hard for a provider which uses multiple task types to implement this function, as it's hard to reason about what shape of data is returned from each task.

Therefore, I feel that we should enforce some kind of task data envelope that all tasks must return from their `makeTestSteps()` or `test()` methods.

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+                console.log(`Error caught in TestBase.execute: ${e}`)
+                console.log(result)
+                return Promise.resolve<unknown>('Replace me!')
+            })
```
In insighst.js we intentionally swallow any errors to be a good third-party citizen and not let let bubble up to the host page. However similar to the debug logging story, I've always wanted to to make it configurable that errors could get reported back to a provider endpoint. Something to discuss/consider.

jakewan replied:
I've removed the console logging from this handler, but it still doesn't do anything useful besides mark the Test object as failed. I'll leave this for future work.

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+     * Convert the beacon data into an encoded payload suitable for transmission
+     * @param data
+     */
+    protected encodeBeaconData(data: unknown): string {
```
A thought. Maybe we should expose this a provider hook instead? What is a provider wanted use use a different wire serialisation method?

jakewan replied:
Agreed... https://github.com/cbsinteractive/open-insights/blob/create-provider-hooks/src/lib/test.ts#L61

phamann commented on src/util/clientinfo.ts:
```diff
@@ -0,0 +1,26 @@
+ * Function that performs a unique API query to get client data
+ * Memoizes responses to ensure that we only hit the API once
+ */
+export function getClientInfo(url: string, resultFun: ClientInfoResponseFunc): Promise<unknown> {
```
`s/resultFun/resultFunc/` ?

jakewan replied:
Yep, done.

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * This is the logic function for conducting an individual test.
+     */
+    execute(): Promise<unknown> {
```
It would be good to get rid of this `unknown` which would clean up a lot of other `unknown`'s across the library. Happily to discuss some ideas.

jakewan replied:
The associated interfaces are now...

```typescript
// @types/index.ts
export interface Executable {
    execute(): Promise<Beacon.Data>
}
```

```typescript
// lib/beacon.ts
export enum State {
    Success,
    Failure,
    Unknown,
}

export interface Data {
    state: State
    testType: string
    data?: {
        [key: string]: any
    }
}

```

phamann commented on src/lib/testBase.ts:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * A subclass implements this method in order to define its specialized mechanics
+     */
+    abstract makeTestSteps(): Promise<unknown[]>
```
I don't understand the difference/relationship between `makeTestSteps()` and `test()`? Isn't having an abstract `test()` a method sufficient?

jakewan replied:
Yep, `makeTestSteps()` is now gone.

phamann reviewed (commented):
Looking good! My comments are a mixture of:
- Add some historic context and question if still needed
- TypeScript/JavaScript style conventions (really not blocking, just would like us to define some) 
- Questions and discussion starters

Hope that helps.

One big thing I'd like us to discuss in the ClientInfo implementations, so will take that to Slack/Call.

jakewan committed:
Interface name consistency and simplification

ProviderInterface -> Provider
IExecutable -> Executable

jakewan committed:
Replace interfaces with function type literals

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Remove extra parentheses

jakewan committed:
Move packages into devDependencies

jakewan committed:
Snake-case filenames for consistency

jakewan committed:
Remove superfluous wrapping promise

jakewan committed:
Make ProviderBase::shouldRun abstract

Feature detection is an important way to avoid breakage in unsupported
browsers. Making this method abstract forces providers to take this into
consideration.

jakewan committed:
More general fetchObject method to replace fetchObjectAndId

Also eliminate console spam. Will consider better logging system.

jakewan committed:
Enable declaration file generation

jakewan committed:
Rename ABC TestBase to Test

jakewan committed:
Implement Test state tracking

jakewan committed:
Port resource timing bug fix from insights.js

Does not include tests. Those will be ported as part of the wider
unit test porting effort.

jakewan committed:
Increase typing of fetch handling variables

jakewan committed:
Increase typing of client info request

jakewan committed:
Added stronger typing to beacon data-making

Also added a hook to allow providers to define their own Resource Timing
validation predicate. Can be generated when Fetch object is created in
case the provider needs different rules based on the type of test being
performed.

jakewan committed:
Consistent function variable naming

jakewan committed:
Improve test setup handling

jakewan committed:
Add further type definitions

- Further type definitions involving generation of beacon data and
  recording of results.

- Rework loadWhenDocumentReady using promises to enable the tag owner to
  process the results when the session is complete.

jakewan committed:
Expose default export from index.ts

jakewan committed:
Use URL class for test resources

Also eliminate a few more unknown type references.

jakewan committed:
Eliminate remaining unknown type references

jakewan committed:
Add fetch support for sending request headers

jakewan committed:
Work on test setup and teardown hooks

jakewan committed:
Add typedoc

jakewan committed:
Add eslint

jakewan committed:
Add prettier

jakewan committed:
Update README.md

jakewan committed:
Address eslint errors

jakewan committed:
Remove unneeded ClientInfoResponseFunc type

jakewan committed:
Work on docblocks

jakewan committed:
Work on dockblocks

jakewan committed:
Work on docblocks

jakewan committed:
Doc SimpleObject interface

jakewan committed:
Complete initial docblocks for src/index.ts

jakewan committed:
Docblock for compose

jakewan committed:
Docblocks for src/util/object.ts

jakewan committed:
Docblocks for clientInfo.ts

jakewan committed:
Docblock defaultSequenceFunc.ts

jakewan committed:
Docblocks for client settings

jakewan committed:
Docblocks for beacon.ts

jakewan committed:
Docblocks for util/beacon.ts

jakewan committed:
Docblock for camelCaseToSnakeCase.ts

jakewan committed:
Docblocks for Test and Fetch

Also eliminate unused Test.makeBeaconURL method

jakewan committed:
Docblocks for providerBase.ts

jakewan committed:
Refactor SessionConfig interface; work on docblocks

jakewan committed:
Refactor TestSetupResult interface; work on docblocks

jakewan committed:
Rename Result to TestResult; work on docblocks

jakewan committed:
Finish TODO docblocks in @types/index.ts

jakewan committed:
Finish docblocks in resourceTiming.ts

phamann commented on package.json:
```diff
@@ -6,17 +6,29 @@
-  },
-  "dependencies": {
-    "@fastly/performance-observer-polyfill": "^1.1.1",
-    "unfetch": "^4.1.0"
```
These should remain as a dep and not a devDep.

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+) => boolean
+
+/**
+ * Represents the result of one aspect of a test. There may be ore than one
```
```suggestion
 * Represents the result of one aspect of a test. There may be more than one
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * A hook enabling a provider to draw a set of zero or more
+     * {@link HttpHeader} tuples from a provider-defined test configuration.
```
This was slightly confusing to me. Maybe:

```suggestion
     * A hook enabling a provider to define a list of zero or more request
     * {@link HttpHeader} tuples to be sent with the test, given a 
     * provider-defined test configuration.
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+
+    /**
+     * A hook enabling a provider to perform encoding of beacon data before
+     * sending it.
```
Add and example of typical encoding:

```suggestion
     * A hook enabling a provider to perform encoding of beacon data before
     * sending it. Such as `JSON.stringify()` or other serialization methods. 
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+     * @remarks
+     * Defaults to {@link defaultSequenceFunc}
+     */
+    sequence: PromiseSequenceFunc
```
I don't know if `sequence` is the correct name for this function now that we're allowing the tag owner to provide a function to handle this (which is a great idea!). The original implementation was called `PromiseSequence` as it processed a list of promises "sequentially" one at a time, where as a site owner could provide a function that ran them all at once or in a different order etc. 

Maybe "run", "runner", "process" or "reduce(r)" or ???

phamann commented on src/index.ts:
```diff
@@ -1,44 +1,89 @@
+ *
+ * @param settings The settings object passed to {@link init}.
+ */
+function startLater(settings: ClientSettings): Promise<SessionResult> {
```
While I know that the value is contained within the settings object, it feels weird to me that this function signature doesn't explicitly contain the delay value. Maybe:

```suggestion
function startLater(delay: number, settings: ClientSettings): Promise<SessionResult> {
```

phamann commented on src/index.ts:
```diff
@@ -1,44 +1,89 @@
+ * @param settings The settings object passed to {@link init}.
+ */
+function start(settings: ClientSettings): Promise<SessionResult> {
+    return Promise.all(
```
> _you're only as strong as your weakest link_

We may have a problem using `Promise.all()` in this context. This will mean that all provider `p.fetchSessionConfig()` will need to be successful for this session to run. Therefore, we're creating a tight coupling between the availability of each providers config service - if fetching over the network etc.

Maybe we could use [`Promise.allSettled()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) instead? This allows us to block until all have completed, regardless of their status, and then filter the results by their status. I.e.

```js
return Promise.allSettled(
    settings.providers
            .filter((provider) => provider.shouldRun())
            .map((provider) => provider.fetchSessionConfig()),
    ).then((sessionConfigResults) => {
        sessionConfigs = sessionConfigResults
            .filter((r) => r.status == "fulfilled")
            .map((r) => r.value)

        sessionConfigs.forEach((v, i) => {
                settings.providers[i].setSessionConfig(v)
                v.executables = settings.providers[i].expandTasks()
        })

        return settings.sequence(sessionConfigs)
    })
```

phamann commented on src/lib/beacon.ts:
```diff
@@ -0,0 +1,42 @@
+     */
+    data?: {
+        /* eslint-disable @typescript-eslint/no-explicit-any */
+        [key: string]: any
```
Could this have the type of `SimpleObject` instead of `any`?

jakewan replied:
This is one I've been struggling with a bit, because `SimpleObject` is limited to primitive types (minus Symbol). I'll poke at it some more today and see if I can find a reasonable compromise. Let me know if you have ideas about this.

jakewan replied:
I think I got it. I extended SimpleObject a bit and added a test to show it being JSON serializable.

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,105 @@
+         */
+        private _isValidEntryFunc: ResourceTimingEntryValidationPredicate = (
+            e,
+        ) => e.requestStart !== 0 && e.connectStart !== e.connectEnd,
```
As a reader, I found this had to parse. Maybe we could move the default predicate value to a const in this file?

```js
const defaultIsValidEntryFunc : ResourceTimingEntryValidationPredicate = (e) => e.requestStart !== 0 && e.connectStart !== e.connectEnd;

// ... later in class definition

private _isValidEntryFunc: ResourceTimingEntryValidationPredicate = defaultIsValidEntryFunc
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+ */
+export interface Executable {
+    /**
+     * Execute a test and return a promise resolving to a test result bundle.
```
```suggestion
     * Execute a test and return a Promise resolving to a test result bundle.
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,375 @@
+     * A hook called very early by the core module, enabling providers an
+     * opportunity to determine if they should participate in the session,
+     * e.g. based on user agent feature detection, random downsampling, etc.
+     * It returns a promise, in case a provider would like to make a network
```
```suggestion
     * It returns a Promise, in case a provider would like to make a network
```

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,105 @@
+            this.fetchObject(),
+            asyncGetEntry(
+                this.getResourceUrl().href,
+                5000,
```
Can we move this to an appropriately named const declaration or a private static member of this class, to make it easier for readers.

```js
const defaultTimeout = 5000;

// ...

asyncGetEntry(this.getResourceUrl(), defaultTimeout, this._ValidEntryFunc)
```

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,105 @@
+        ]).then(
+            ([response, entry]): Promise<TestResultBundle> => {
+                return this._provider.createFetchTestResult(
+                    normalizeEntry(entry),
```
Non blocking comment / question: will all providers want the entry to be normalized per our spec?

jakewan replied:
That's a good point, similar to the comment below regarding `EXCLUDED_PROPS`. It probably makes more sense to pass the unmodified entry into `provider.createFetchTestResult` and let it do what it wants.

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,105 @@
+        const requestHeaders: [
+            string,
+            string,
+        ][] = this._provider.getResourceRequestHeaders(this._config)
```
Why doesn't this have the type `HttpHeader[]`?
```suggestion
        const requestHeaders: HttpHeader[] = this._provider.getResourceRequestHeaders(this._config)
```

jakewan replied:
I just removed the type specifier, since it's implied.

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,105 @@
+                    return accumulator
+                },
+                {},
+            )
```
While I think this reduce is fine. We could simplify this logic if  `provider.getResourceRequestHeaders()` returned a string/string dictionary object instead of a list of tuples. Then we could pass it straight to `init.headers` and avoid the reduce mapping from one data structure to the other.

jakewan replied:
Makes sense. The case where headers with the same name might be sent multiple times should be rare and can be the responsibility of the provider to manage.

phamann commented on src/lib/resourceTiming.ts:
```diff
@@ -1,110 +1,158 @@
+ * A list of Resource Timing entry properties not to include in the normalized
+ * transformation used internally.
+ */
+const EXCLUDED_PROPS = ["name", "initiatorType", "entryType"]
```
Non blocking comment / note: this list was a Fastly implementation detail. Do we think every provider will mind having these properties removed?

jakewan replied:
This was carried along into the <provider>/src/lib/normalizeEntry.ts modules.

phamann commented on src/lib/resourceTiming.ts:
```diff
@@ -1,110 +1,158 @@
+ * resolves with the entry once successful. If the timeout is reached before
+ * and entry is returned it rejects the Promise.
+ * @param name URL of the of the entry
+ * @param timeout Optional timeout
```
Is it optional?

jakewan replied:
Nope, not optional. There was a default, but managing it was moved upstream in response to an earlier comment.

phamann commented on src/lib/getNetworkInformation.ts:
```diff
@@ -0,0 +1,23 @@
+export default function getNetworkInformation(): NetworkInformation {
+    const navWithInfo = navigator as Navigator
+    const connection = navWithInfo && navWithInfo.connection
+    if (connection) {
```
We can use optional chaining here, which was introduced in TypeScript 3.7:
```suggestion
    const connection = navWithInfo?.connection
    if (connection) {
```

phamann commented on src/lib/providerBase.ts:
```diff
@@ -0,0 +1,151 @@
+     * @remarks
+     * Provides a simple mechanism using the Beacon API, with the Fetch API as
+     * a fallback. Providers may override this if they require more specialized
+     * reporting, e.g. utilize websockets, invoke a 3rd party library, etc.
```
```suggestion
     * reporting, e.g. utilize WebSockets, invoke a 3rd party library, etc.
```

phamann commented on src/lib/test.ts:
```diff
@@ -0,0 +1,116 @@
+                    this._config,
+                    bundle,
+                )
+                this._provider.sendBeacon(
```
This returns a `Promise` which is being swallowed here and should be returned form this function to ensure control flow is maintained.

jakewan replied:
@phamann I added resolve and reject handlers for this Promise and also added a nascent error handling hook. Could you look closely at this to see if it's like what you had in mind?

phamann commented on src/util/beacon.ts:
```diff
@@ -1,17 +1,29 @@
+ *
+ * @remarks
+ * Uses the Beacon API if available. Otherwise, sends a POST message using the
+ * Fetch API.
```
Consider linking to the Beacon and Fetch API docs on MDN.

phamann commented on src/util/clientInfo.ts:
```diff
@@ -0,0 +1,44 @@
+ * Type capable of mapping URLs (as strings) to Promise<{@link ClientInfo}>
+ * objects.
+ */
+interface Cache {
```
Should this be a Type and not an Interface?

jakewan replied:
Is this question referring to Cache's definition as an interface or to the fact that the comment refers to it as a "Type"? I've been somewhat unsure when to use one over the other, and have kind of blindly been preferring interfaces based on what I read here:

https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases

>Because an ideal property of software is being open to extension, you should always use an interface over a type alias if possible.
>
>On the other hand, if you can’t express some shape with an interface and you need to use a union or tuple type, type aliases are usually the way to go.

Since it didn't seem like Cache would need to be used in union or tuple definitions, I went with an interface.

In case the question was more about the terminology used in the docblock, I'll update that now.

phamann commented on src/index.ts:
```diff
@@ -1,44 +1,89 @@
+ * settingsBuilder.addProvider(new Provider(fooSettings))
+ *
+ * // Execute a RUM session
+ * init(settingsBuilder.value)
```
```suggestion
 * init(settingsBuilder.toSettings())
```

phamann commented on src/util/loadWhenDocumentReady.ts:
```diff
@@ -1,15 +1,17 @@
-  // Otherwise, attach event listener to invoke when ready
-  document.addEventListener("readystatechange", (): void => {
+/**
+ * Utility to resolve a promise when the browsers ready state reaches "complete"
```
```suggestion
 * Utility to resolve a Promise when the browsers ready state reaches "complete"
```

phamann commented on src/util/promiseSequence.ts:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Provides a mechanism to create a list of promises one at a time.
```
```suggestion
 * Provides a mechanism to invoke a list of promises one at a time.
 * This can used by a {@link PromiseSequenceFunc} instead of Promise.all
```

phamann commented on webpack.common.js:
```diff
@@ -1,23 +1,23 @@
-    },
+        filename: "main.js",
+        path: path.resolve(__dirname, "dist"),
+        library: "openInsights",
```
Maybe add a //TODO comment here that this may need to change depending on what we decide to call this project.

phamann commented on package.json:
```diff
@@ -6,17 +6,29 @@
+    "ts-jest": "^26.1.0",
+    "ts-loader": "^7.0.5",
+    "typedoc": "^0.17.7",
+    "typescript": "^3.9.3",
```
Nice!

phamann reviewed (changes requested):
@jakewan This is looking GREAT! Thank you so much for your hard work so far. The addition of the typedoc comments made this review much nicer and easier for me to grok (especially being away from the project for some weeks).

I've left a mixture of, typos, questions and style/readability improvements. Mostly non-blocking, however the two major ones are:

- The use of  `Promise.all()` in the main `start()` function
- Swallowing of the `sendBeacon` Promise in the main `test.execute` method.

Outstanding questions:
- Tests?
- When ready, this PR should probably squashed into a single "initial commit" (or similar) before merging. And if ok with you myself and any others you think appropriate being added as co-authors.

jakewan commented:
@phamann 
>When ready, this PR should probably squashed into a single "initial commit" (or similar) before merging. And if ok with you myself and any others you think appropriate being added as co-authors.

Oh yes, absolutely.

jakewan committed:
Fix dependencies

@fastly/performance-observer-polyfill and unfetch need to be
dependencies and not development-only dependencies.

jakewan committed:
Update src/@types/index.ts

Fix typo

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update src/@types/index.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Rename ClientSettings.sequence to ClientSettings.sessionProcess

Also renamed various other related symbols.

jakewan committed:
Update docblock for Provider.encodeBeaconData

jakewan committed:
Fix `init` example in src/index.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update src/util/loadWhenDocumentReady.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update src/util/promiseSequence.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Add TODO comment concerning name of the global library object

jakewan committed:
Fix whitespace

jakewan committed:
Update src/@types/index.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Rebuild docs

jakewan committed:
Add delay parameter to `startLater()`

jakewan committed:
Use Promise.allSettled

jakewan committed:
Add provider name to TestResultBundle

jakewan committed:
Revert "Add provider name to TestResultBundle"

This reverts commit 4096497f41f785dc2806b39f7cfb545499bae92c.

jakewan committed:
Add optional providerName to TestResultBundle

jakewan committed:
Rebuild docs

jakewan committed:
Add some unit tests (83% coverage)

jakewan committed:
Update src/@types/index.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update src/lib/providerBase.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update comment in Test exception handler

jakewan committed:
Update src/lib/getNetworkInformation.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Refactor Fetch's Resource Timing validation function member definition

jakewan committed:
Fetch has default timeout for call to asyncGetEntry

jakewan committed:
Resource Timing normalization made responsibility of provider

jakewan committed:
Include MDN doc links in docblock

jakewan committed:
Update clientInfo.Cache docblock

jakewan committed:
Update npm dependencies

jakewan committed:
Build docs after updating typedoc

jakewan committed:
Handle Promise returned from Provider.sendBeacon

Also introduces a Provider.handleError hook that providers can use to
process errors emitted by the core module.

jakewan committed:
Supply reject handler

jakewan committed:
Update type comments

jakewan committed:
Simplify return value of Provider.getResourceRequestHeaders

jakewan committed:
Extend SimpleObject

Includes:

- Use SimpleObject rather than any as Beacon.Data.data type.

- Fix broken unit tests.

jakewan committed:
Set to version 0.0.1

phamann commented on package.json:
```diff
@@ -1,19 +1,37 @@
 {
   "name": "open-insights",
-  "version": "1.0.0",
+  "version": "0.0.1",
```
Nit: this should be `0.1.0`
```suggestion
  "version": "0.1.0",
```

phamann commented on package.json:
```diff
@@ -1,19 +1,37 @@
-    "build:dev:watch": "webpack --config webpack.dev.js --watch"
+    "build:dev:watch": "webpack --config webpack.dev.js --watch",
+    "build:doc": "rimraf doc && typedoc",
+    "lint": "eslint --fix src",
```
I find it useful to also run the TypeScript compiler `tsc` in this command to catch any obvious errors preventing compilation.
```suggestion
    "lint": "tsc --noEmit && eslint --fix src",
```

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,415 @@
+ *
+ * Alias for {@link SimpleObject}
+ */
+export type ResourceTimingEntry = SimpleObject
```
Note: I've battled with this myself in `insights.js` and `performance-observer-polyfill`. Ideally, I'd prefer to use an offical type exported by the TypeScript DOM library, such as `PerformanceEntry` but it's never been sufficient for me and the `toJSON()` method always gets in the way.

From the TypeScript source: https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L11506-L11512  
```
interface PerformanceEntry {
    readonly duration: number;
    readonly entryType: string;
    readonly name: string;
    readonly startTime: number;
    toJSON(): any;
}
``` 

Maybe we could try using one of the fancy utility types, such as [`Omit<T, K>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittk) here? Such as:
```suggestion
export type ResourceTimingEntry = Omit<PerformanceEntry, 'toJSON'>
```
I haven't tried this so don't know if it will work. 🤣

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,415 @@
+     * @param errorType
+     * @param innerError
+     */
+    handleError(errorType: KnownErrors, innerError: Error): void
```
Nice 🔥

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,415 @@
+     * @param setupResult The provider-defined result of any test setup
+     * activity done prior to the fetch.
+     */
+    createFetchTestResult(
+        timingEntry: ResourceTimingEntry,
+        response: Response,
+        testConfig: TestConfiguration,
+        setupResult: TestSetupResult,
+    ): Promise<TestResultBundle>
```
This is much cleaner than before. Very flexible 👌

phamann commented on src/@types/index.ts:
```diff
@@ -0,0 +1,415 @@
+     * @remarks
+     * Defaults to {@link defaultSequenceFunc}
+     */
+    sessionProcess: SessionProcessFunc
```
If this has a default set elsewhere in the library, should it be optional?
```suggestion
    sessionProcess?: SessionProcessFunc
```

jakewan replied:
This is a good point. Right now there is a default, but it's not in a good place, `ClientSettingsBuilder`, and there's no guarantee users will necessarily even use this class. It seems more intuitive for it to be optional here, and introduce the default function downstream if needed.

phamann commented on src/index.ts:
```diff
@@ -1,44 +1,99 @@
-        console.log('No providers promised to make config calls')
-    }
+/**
+ * Called immediately by {@link init} if no
+ * {@link ClientSettings.preConfigStartDelay} setting has been specified.
```
Note: There is a lot happening in this function, we should consider fleshing out this description to explain what it does a bit more. Something like:
```suggestion
 * Starts the RUM session by consuming all provider configurations and converting them to 
 * tasks using the following logic: and expanding them into tasks 
 * 
 * 1) Asynchronously fetches each provider session config and wait for them all to settle.
 * 2) Filter provider config results to only those which fulfilled successfully.
 * 3) For each provider set the configuration and expand the tasks.
 * 4) Process the expanded tasks.
 *  
 *Called immediately by {@link init} if no
 * {@link ClientSettings.preConfigStartDelay} setting has been specified.
```

phamann commented on src/index.ts:
```diff
@@ -1,44 +1,99 @@
+                (r) => (r as PromiseFulfilledResult<ExecutableContainer>).value,
+            )
+        sessionConfigs.forEach((v, i) => {
+            const p = settings.providers[i]
```
Unfortunately, looking up the provider by index in the static `ClientSettings.providers` array is potentially brittle. 

We are first filtering the providers by their `.shouldRun()` predicate and then by whether or not their `fetchSessionConfig()` fulfilled successfully. Therefore,  we may end up with a list of `sessionConfigs` smaller than the original provider list. 

So we'll have to maintain another lookup internal to this function. Something like:
```typescript
function start(settings: ClientSettings): Promise<SessionResult> {
    let providers = settings.providers
        .filter((provider) => provider.shouldRun())
        
    return Promise.allSettled(
        providers.map((provider) => provider.fetchSessionConfig())
    ).then((settled) => {
        providers = providers.filter((p, i) => settled[i].status === "fulfilled")
        const sessionConfigs = settled
            .filter((r) => r.status === "fulfilled")
            .map(
                (r) => (r as PromiseFulfilledResult<ExecutableContainer>).value,
            )
        sessionConfigs.forEach((v, i) => {
            const p = providers[i]
            p.setSessionConfig(v)
            v.executables = p.expandTasks()
        })
        return settings.sessionProcess(sessionConfigs)
    })
}
```

phamann replied:
This also makes me think we should add some tests to cover these edge cases, such as a failing`fetchSessionConfig` etc.

jakewan replied:
Great catch!

phamann commented on src/lib/beacon.ts:
```diff
@@ -0,0 +1,38 @@
+import { TestConfiguration, SimpleObject } from "../@types"
```
Is there are reason why these types aren't located with all the others in `src/@types/index.ts`?

jakewan replied:
Haha, no reason. I had just learned about modules as namespaces and was probably too eager to try it. I'll move them.

phamann commented on src/lib/errors.ts:
```diff
@@ -0,0 +1,6 @@
+     * An error occurred when attempting to send beacon data.
+     */
+    SendBeacon,
+}
```
This is great! I can think of couple of other "known" errors, which we should try and cover:
- Provider config fetch error 
- Test resource fetch error
- Test executution

phamann commented on tsconfig.json:
```diff
@@ -3,12 +3,16 @@
     "strict": true,
     "esModuleInterop": true,
-    "forceConsistentCasingInFileNames": true
+    "forceConsistentCasingInFileNames": true,
+    "declaration": true,
+    "declarationDir": "dist/types",
+    "declarationMap": true
```
I'd be interested in seeing how much work it would be to turn on some of TypeScripts "good parts" here. Such as `noImplicitAny` and `strictNullChecks`. But this is definitely something that _can_ and _should_ be applied in a later PR not this. One, however commenting now for notes and discussion :)

jakewan replied:
It's my understanding that having `"strict": true` basically enables most such settings. Is that not the case?

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Class representing a basic "fetch" test. Along with its parent class
+ * {@link Test}, the class provides a number of hooks enabling providers to
+ * control certain implementation details supporting their use case.
```
Added some more detail to docs as this is the main test type most providers will use.
```suggestion
 * Class representing a basic "fetch" test. It fetches a supplied resource from the 
 * network and collects the corresponding ResourceTimingEntry from the 
 * Performance timeline. 
 *
 * Along with its parent class {@link Test}, the class provides a number of hooks 
 * enabling providers to control certain implementation details supporting their 
 * use case.
```

phamann commented on src/lib/fetch.ts:
```diff
@@ -0,0 +1,110 @@
+     * result of calling {@link Provider.createFetchTestResult} when the test
+     * data has been obtained.
+     */
+    test(setupResult: TestSetupResult): Promise<TestResultBundle> {
+        const defaultTimeout = 5000
+        return Promise.all<Response, ResourceTimingEntry>([
+            this.fetchObject(),
+            asyncGetEntry(
+                this.getResourceUrl().href,
+                (this._config as FetchConfiguration)
+                    .performanceTimingObserverTimeout || defaultTimeout,
+                this._isValidEntryFunc,
+            ),
+        ]).then(
+            ([response, entry]): Promise<TestResultBundle> => {
+                return this._provider.createFetchTestResult(
+                    entry,
+                    response,
+                    this._config,
+                    setupResult,
+                )
+            },
+        )
+    }
```
Nice! Really like how this has turned out.

phamann commented on src/lib/providerBase.ts:
```diff
@@ -0,0 +1,185 @@
+     * See {@link Provider.handleError}
+     */
+    handleError(errorType: KnownErrors, innerError: Error): void {
+        throw new Error("Providers should override Provider.handleError")
```
Do we really want this to throw even with the helpful message? If we want them to override then we should make this abstract or a no-op.

phamann commented on src/util/beacon.ts:
```diff
@@ -1,17 +1,43 @@
+        if (navigator.sendBeacon(url, data)) {
+            return Promise.resolve()
+        }
+        // TODO: should we attempt to use fetch in this case, or does that risk
+        // duplicate reporting?
```
Great question. Looking at the [spec processing model](https://www.w3.org/TR/beacon/#h-sec-processing-model) it seems pretty safe for us to fall through this block to the fetch below. And something i've not considered before 👍

phamann commented on src/util/clientInfo.ts:
```diff
@@ -0,0 +1,43 @@
+}
+
+/**
+ * Function that performs a unique API query to get client data
```
```suggestion
 * Function that performs a unique API query to get client data.
```

phamann commented on src/util/clientInfo.ts:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Cleans the cache: necessary for testing
+ * @param url URL and its associated Promise<{@link ClientInfo}> to remove from
+ * the cache
```
```suggestion
 * the cache.
```

phamann commented on src/lib/test.ts:
```diff
@@ -0,0 +1,132 @@
+        this._state = TestState.Running
+        return this._provider
+            .testSetUp(this._config)
+            .then((setupResult) => this.test(setupResult))
```
We should consider `catch`'ing here and calling something like `onSendTestRejected()`

phamann reviewed (commented):
This is looking **AMAZING**! Really exciting to see how it's evolved over the last month. 

My comments are mostly non-blocking type related questions, some documentation additions and some style changes. **I'd say the only blocker is the refactor of the `start()` function to fix the bug I noticed.**

Couple of things to discuss offline and shouldn't block this PR being merged:
- testing: I haven't really dug into the testing patterns but they all seemed well thought out and ok to read. I have one question around `sinon` it seems you're mostly using it for `stub` and `sandbox`. The "jest way" is that you shouldn't need to reach out for other libraries and you should be able to use their spies/mocks over sinon stubs and seems like [jest-sandbox](https://github.com/JoinColony/jest-sandbox) covers the sandboxing needs. Again, not a blocker as I don't have any strongly held opinions here :)
- debug logging
- error reporting: there are a couple more places in the library we can decorate with `onReject` style hooks, or simply call the providers `onError` handler. The patterns you've introduced are already a big way towards hooking up a provider supplied error reporting endpoint.
- `tsc` options via `tsconfig.json`
- We should enable dependabot on the final repo location to make sure we get automated patches for dependency upgrades.
- Integration test style server via `npm run start` or similar which loads a html page containing a basic provider and single task etc. Which I often find useful for debugging.

jakewan committed:
Update package.json

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update package.json

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Update version in package-lock.json

jakewan committed:
Abstract ProviderBase.handleError

jakewan committed:
Add UnitTestProvider.handleError

jakewan committed:
Typo in src/util/clientInfo.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Typo in src/util/clientInfo.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan commented:
>testing: I haven't really dug into the testing patterns but they all seemed well thought out and ok to read. I have one question around sinon it seems you're mostly using it for stub and sandbox. The "jest way" is that you shouldn't need to reach out for other libraries and you should be able to use their spies/mocks over sinon stubs and seems like jest-sandbox covers the sandboxing needs. Again, not a blocker as I don't have any strongly held opinions here :)

Definitely would prefer to minimize dependencies and stick to the "jest way" as much as possible. Actually the main thing that influenced me to use sinon here, aside from past familiarity, was the `createStubInstance` function. However, that was mainly laziness on my part. There's probably a jest way to do that too that I need to learn.

jakewan committed:
Beacon.Data -> BeaconData; Beacon.State -> BeaconState

jakewan committed:
Make ClientSettings.sessionProcess optional

jakewan committed:
Fix broken unit tests and lint errors

jakewan committed:
Additional error types

jakewan committed:
Update src/lib/fetch.ts

Co-authored-by: Patrick Hamann <patrick@fastly.com>

jakewan committed:
Remove trailing space

jakewan committed:
Refactor src/index.test.ts

jakewan committed:
Handle edge cases in start()

Includes:

* Maintain internal lookup of providers that should run

* Add unit test covering when one or more provider doesn't run

* Add unit test covering when one or more providers' session config
request fails.

Commit:
Add a GitHub Actions workflow for pull request builds. (#5)
